### PR TITLE
[dune subst] Improve dune-project (name ..) behavior

### PR DESCRIPTION
**Pull request by rgrinberg**
This PR re-organizes the `dune subst` tests, and adds a test case for what I
think is strange behavior. Why should `dune subst` impose additional
restrictions that dune itself doesn't check for?

### CODE REVIEW

**Comment by jeremiedimino** [comment]
It's purely technical: we need a package definition to fill in some of the substituted var. For that, we take the package description that matches the project name. Maybe we can do otherwise, in which case it is fine to lift this restriction.

**Review by jeremiedimino — APPROVED** [review]

**Comment by rgrinberg** [comment]
I see. Tbh, it's a bit confusing to bless a particular package this way. I suppose we could fix it this way:

Allow `maintainer`, `authors`, `homepage`, etc as toplevel fields in the dune-project files. Effectively, this will allow us to attach this metadata to the project rather than the package. We could also use this project metadata to fill in as default for the packages in the project. E.g.:

```
(homepage foobar.com)
(package (name p1))
(package (name p2))
```

Both p1 & p2 will inherit the homepage field.

Things will still be a bit confusing as `PKG_*` fields will refer to project rather than package metadata, but we really don't have a way to make `$ dune subst` package rather than project based anyway.

If we don't fix it as i've pointed out above, we should at  least consider adding a warning for this. Dune should warn as early as possible that this project isn't compatible with `dune subst`.

**Comment by emillon** [comment]
`maintainers`, `authors` and `source` (I assume `homepage` etc too) are already available at the toplevel, so that part can be already made to work I think.

In addition, we can also relax the check so that it's only an error if `PKG_*` are actually used as watermarks.

**Comment by rgrinberg** [comment]
Here's a proposal:

If we generate the opam files, we are going to use the project's information for some substitutions. If not, we'll stick to told behavior of looking for a package that match's the project's name. To be on the safe side, we could perhaps make it 2.8 only, but this seems like a safe way to improve the behavior.

**Comment by emillon** [comment]
Seems good but does that mean that (name) is optional in that case?

**Comment by rgrinberg** [comment]
@emillon it would still be required like it is now. The only difference is that it doesn't need to match any package name.

**Comment by emillon** [comment]
OK, that sounds great then.

**Comment by rgrinberg** [comment]
Okay let's see if @jeremiedimino approves.

**Comment by jeremiedimino** [comment]
SGTM. However, we do need to continue generating the same opam files for existing projects. So the new behaviour should be for >=2.8 projects only.

**Comment by jeremiedimino** [comment]
And in the future, maybe we should change the set of variables `dune subst` rewrites. It seems that the `PKG_...` ones don't make much sense anymore

**Comment by rgrinberg** [comment]
Note that opam file generation isn't changed in this PR. Only the behaviour of dune subst.

**Comment by rgrinberg** [comment]
A related fix that I'd like to consider:

If a user is generating an opam file through dune, we should warn them if there's no project name. The reason for this is that we generate instructions that include `dune subst`. This means that we should include the project name for dune subst to work reliably.

**Comment by emillon** [comment]
I think it's a good idea. The experience with `dune subst` failing is not great because you don't experience it until you `opam pin` days later, and then you don't have context anymore.

**Comment by jeremiedimino** [comment]
> Note that opam file generation isn't changed in this PR. Only the behaviour of dune subst.

Agreed. However, `dune-release` calls `dune subst`, and it doesn't feel right that the tarball you get depends on the version of the dune binary installed rather than the source only.

The other change you are proposing seems good to me.

**Comment by rgrinberg** [comment]
@emillon @jeremiedimino I improved the behavior as proposed:

* I emit a warning when a user is missing `(name ..)`. Limited to only when dune is being used to generate the opam files. This way we can be reasonably confident the intention is to use dune subst.
* I preserve the old behavior for lang dune < 2.8

**Comment by rgrinberg** [comment]
@emillon can I assign this to you for review? I'd like to get this in 2.8 along with the other dune subst fixes.

**Comment by rgrinberg** [comment]
ping @jeremiedimino or @emillon this needs another review since the implementation changed since the last approval

File: src/dune_rules/watermarks.ml
Comment (jeremiedimino) [inline src/dune_rules/watermarks.ml:150]:
```suggestion
      (* dir_status only affects warning status, but it will not matter
```

Comment (jeremiedimino) [reply src/dune_rules/watermarks.ml:150]: To be fair, in this case the code is clear enough with a comment; `Normal` is the value I would expect in `dune subst`. It would be weird for `dune subst` to pass any other value here

**Comment by jeremiedimino** [comment]
I read the changes. One observation: for lang < 2.8, we are now less strict. Up to now, if there was no package matching the project name, we were failing. After this PR, we fallback to the project info.

Another observation is that technically, dune subst could produce different results after bumping the `lang` line. For instance, if the user has `opam.template` files, or overrides some of the fields in the `(package ...)` stanza. Finally, thinking about it again I find that tying the behavior of `dune subst` - a general purpose feature - to `(generate_opam_files ...)` - an opam specific feature - a bit odd.

Here is a simpler proposal:

- `lang < 2.8`: we change nothing. Not having a package matching the project name is a hard error
- `lang >= 2.8`: we still look for a corresponding package, but we fallback to the project info otherwise.  Put another way, we only relax the criteria but with no material change
- `lang >= 3.0.0`: we only look at the project info, and we also rename the set of variables to keep things consistent

**Comment by rgrinberg** [comment]
> I read the changes. One observation: for lang < 2.8, we are now less strict. Up to now, if there was no package matching the project name, we were failing. After this PR, we fallback to the project info.

That seems like a strict improvement to me. There are now some users running into this at #3965 as well.

> Another observation is that technically, dune subst could produce different results after bumping the lang line. For instance, if the user has opam.template files, or overrides some of the fields in the (package ...) stanza

That also seems desirable as we are basically fixing a bug for free for them. It seems like a bug to rely on metadata that comes from a package rather than a project.

> Finally, thinking about it again I find that tying the behavior of dune subst - a general purpose feature - to (generate_opam_files ...) - an opam specific feature - a bit odd.

Yeah, I also found it a little unsatisfactory. It's all a little problematic because all of this is only a problem for the user only if they're actually using `dune subst`. `(generate_opam_files)` is used here to pick up their intent with a high degree of certainty. Even this isn't perfect because a user could use `(generate_opam_files)` and modify the `.template` file to not use `dune subst`, but this case didn't seem very important to me.

> lang < 2.8: we change nothing. Not having a package matching the project name is a hard error

I'm afraid this will keep biting users then. I see the current behavior is buggy, but not because it's a hard error, but because they only see the error *after* they've released their package.

**Comment by jeremiedimino** [comment]
I see your point, but if a newer dune relaxes the condition for older language versions, then we are breaking our promise to users. For instance, if someone works on a 2.7 project with a 2.8 dune binary, they might see no error. Then, because the package claims to be compatible with Dune >= 2.7, someone else might try to install the package with a dune 2.7 binary and it wouldn't work because dune 2.7 was more strict.

**Comment by rgrinberg** [comment]
This PR is now ready again and I've implemented Jeremie's last suggestion.

File: src/dune_rules/watermarks.ml
Comment (emillon) [inline src/dune_rules/watermarks.ml:336]: It seems unusual to refer to dune lang 3.0 here. Maybe we can just add a `(* DUNE3 *)` comment?

File: src/dune_engine/dune_project.ml
Comment (emillon) [inline src/dune_engine/dune_project.ml:833]:
```suggestion
                  subst works (for example for release builds or when package in pinned by opam)"
```

File: src/dune_rules/watermarks.ml
Comment (emillon) [inline src/dune_rules/watermarks.ml:326]: It surprised me that the message directly refers to (package) stanza rather than the existence of an opam package. To clarify, it's not enough to have a `package.opam` file?

**Review by emillon — COMMENTED** [review]
I think that the docs for this new behavior should be added in doc/usage.rst. Otherwise the solution as described discussion looks good to me!

File: src/dune_rules/watermarks.ml
Comment (rgrinberg) [reply src/dune_rules/watermarks.ml:336]: We do it in other parts of the code as well. If we already have it implemented for the future version, might as well leave it in.

File: src/dune_engine/dune_project.ml
Comment (rgrinberg) [reply src/dune_engine/dune_project.ml:833]: I reworded to mention release builds and pins without the parens

File: src/dune_rules/watermarks.ml
Comment (rgrinberg) [reply src/dune_rules/watermarks.ml:326]: Having a `package.opam` is enough. I'll get rid of the 2nd sentence as adding a package can be done in two ways, and both are simple.

**Pull request merged**
